# Python Implementations and Applications

<div class="alert alert-block alert-warning">
Hey everyone! Today, we'll be going over some more specific examples of common Python applications. Meaning, we'll dive deeper into some more useful and technical packages supported by Python. Just a brief caveat: there may be parts of this talk that are overly technical, or just may not be relevant to you. That's okay. I think it's important to go over examples of more technical Python use-cases, so we have a greater perspective about what Python can be used for. And, as long as you have some Python knowledge, or programming experience, I feel it's okay if you're feeling a little lost at certain parts of this talk, since I essentially divided this talk into segments.

<br />

In the beginning stages of this talk, we'll implement our own NLP model. Specifically, we'll implement a sentiment classifier using a pre-trained neural network called BERT. This model will allow us to make very accurate predictions about the sentiment of movie reviews, which will hopefully be somewhat fun for some of you.

<br />

After implementing our model, we'll demonstrate how we can write some of the important output of our model, such as testing examples, to a database using Python. This will hopefully be cool for us to see how Python can be useful to not only data scientists, but data engineers as well, and how they can all work together.

<br />

Lastly, we'll move into some back-end engineering tasks, like setting up an actual web service using Python. More specifically, we'll want to make our sentiment classifier available to other people for testing purposes. So, we'll use a Python package, called Flask, to create a way for other people access our Python classifier on the web.

<br />

And, we'll also want to create an actual front-end for users to interact with our sentiment classifier web service. So, we'll use another Python package to create an interface, along with some stylish visualizations, so users can play around with it.

<br />

But, for now, let's just focus on building a pretty simple NLP model.
</div>

# Introducing NLP Tasks

<div class="alert alert-block alert-warning">
Before we go any further into model building, let me actually just take a few moments to help motivate sentiment analysis and NLP tasks as a whole.
</div>

### History of NLP

---

<div class="alert alert-block alert-warning">
Before we go any further into model building, let me actually just take a few moments to help motivate sentiment analysis and NLP tasks as a whole. First, let's start off with a quick history lesson of how NLP has developed over the last decade or so.
</div>

- Lots of progress in NLP over the last decade!
- CBOW
    - Used in early stages of NLP
    - Advantage: Predicts the next word in a sentence
    - Disadvantage: Limited to fixed sentences
    - Disadvantage: Excludes relationships in a sentence
- ELMo
    - Created at the start of 2018
    - Advantage: Predicts the next word in a sentence, and more!
    - Advantage: Captures relationships in a sentence
    - Disadvantage: Harder to capture relationships in longer sentences
- BERT
    - Created at the end of 2018
    - Advantage: Predicts the next word in a sentence, and more!
    - Advantage: Captures relationships in a sentence, even in longer sentences!

<div class="alert alert-block alert-warning">
So, let me just start off by saying there's been a lot of progress in NLP over the last decade. Going back to the early 2000s, a guy named Bengio proposed the first neural language model using a neural network, where machine translation was fairly expensive and somewhat hard to solve. At that time, statistical models, such as the continuous bag-of-words model, were used to solve fairly simple NLP tasks, such as next-word prediction and spell checking. This model was useful for predicting next words in sentences, but lacked the ability to pick up on the context of words in sentences, or really understand any of the relationships between words.

<br />

Then, starting at around 2010, many more complex NLP problems were beginning to be solved. For one, Apple's Siri became known as one of the world’s first successful NLP assistants to be used by general consumers. So, problems like question and answering and chat bots were beginning to be solved. And, these types of NLP models were beginning to be able to pick up on the context of words in sentences. But, these tasks were still fairly expensive and slow. Plus, it was still somewhat difficult to understand the relationships between words for longer sentences.

<br />

Then, in the last decade, companies like Google, Facebook, Apple, and others, were able to build more complex neural network architectures, which allowed them to solve more complex NLP problems. And, they were able to capture relationships in very long sentences fairly well. Google, for example, trained a network called BERT, which was able to solve some of these problems very well. Other networks have since been released by other companies, but they really pioneered this movement towards using sequence networks for NLP tasks, which allowed companies to easily create chat bots, summarize text, perform auto-completion, translate text from one language to another, and so much more.
</div>

### Applications for BERT

---

- Next-sentence prediction
- Multi-task language modeling
- Text classification
- Question answering
- Sentiment analysis
- So much more!

<div class="alert alert-block alert-warning">
Yeah, so here, I just want to list some of the applications for BERT. I've only listed a few, like next-sentence prediction, multi-task language modeling, text classification, question answering, and sentiment analysis. But, just know, there are so many more applications solved by NLP networks like BERT and other similar networks, like GPT3.
</div>

### Motivating NLP Tasks

---

<div class="alert alert-block alert-warning">
Now that we've briefly discussed the history and development of some of these important NLP applications, let me quickly touch on how NLP models differ from a lot of generic classification models, including image classification models, that have existed for a while.
</div>

- NLP models have the following properties:
    - Designed for handling sequential data
    - Can allow multiple outputs
- Most NLP data is messy
- Most NLP data preprocessing is cumbersome
- Training accurate NLP models is slow

<div class="alert alert-block alert-warning">
Specifically, NLP models usually are designed for handling sequential data, since most NLP problems only use some series of words as input, where these words represent some sequential series of words in most cases. Also, NLP models can receive a single input or multiple inputs, and they can output a single value or multiple values. In most generic classification models, including image classification models, they only really allow a single output to be spit out.

<br />

Also, a staple of most NLP models is that the data is fairly messy and needs a fair amount of data preprocessing. The training can be very slow as well, compared to simpler, more generic classification models.
</div>

### What is BERT?

---

<div class="alert alert-block alert-warning">
Now that we've introduced the use of neural networks to solve some of the more complex NLP problems, let's briefly motivate the reason for using BERT in particular for some of these NLP problems, and sentiment analysis in particular.
</div>

- BERT is a pre-trained Transformer model
- Pre-trained on:
    - English Wikipedia (2500M words)
    - BookCorpus (800M words)
- Achieved the best accuracies for NLP tasks in 2018

<div class="alert alert-block alert-warning">
Well, for starters, BERT is a pre-trained transformer model. So, over the last decade, Google has spent billions of dollars researching ways of solving some of these complex NLP problems, and developing a neural architecture to do this efficiently. As I mentioned earlier, they were able to do this in 2018 when they created BERT. Generously, they open-sourced this model to the public. Meaning, anyone could use their state of the art model for free.

<br />

And, just to be clear, this wasn't just another NLP model. This was a very robust and accurate model that could be applied to many different NLP tasks. It was also trained on a huge amount of data, that would have been very difficult for any other person or company to process. Specifically, the model learned from all of the English words from Wikipedia, which consisted of 2500M words. On top of that, it also learned from a a dataset consisting of 800M words, which came from published and unpublished books.

<br />

As a result, it was able to achieve the best accuracies for so many NLP tasks when it was released in 2018, such as sentiment analysis, text summarization, and many more. Since then, many other companies and researchers have made their own version of BERT by slightly tweaking their model, such as GPT3 and the T5 model. Some of these models produce even better accuracies as well. However, most of these models have very comparable accuracies to BERT for many NLP tasks. And, some of these models I mentioned haven't even been fully released yet. So, for now, we'll focus on implementing BERT.

<br />

Even though there are some slightly improved models out there, we'll use BERT to build our own sentiment classifier. I'll also show you how you can improve BERT to learn more refined tasks, such as classifying sentiments for a particular topic, such as movie reviews. Let me move on so I can explain this in more detail.
</div>

# Predicting Sentiment of Movie Reviews

<div class="alert alert-block alert-warning">
Now that I've motivated the importance of some of the most major NLP problems seen at various companies, let's go back to actually implementing the BERT model, and specifically building a model that predicts the sentiment of IMDB movie reviews, which will hopefully be fun for us.

<br />

Before we can start building a model, let's first load in the data.
</div>

### Loading IMDB Movie Reviews

---

In [1]:
from simpletransformers.classification import ClassificationModel
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from sklearn.model_selection import train_test_split
import pandas as pd
import tensorflow as tf

<div class="alert alert-block alert-warning">
Here, we're just loading in some packages used for reading in the csv consisting of 50k IMDB reviews. We're also reading in the BERT model that we'll use for predicting the sentiment of a given review, which we'll go over later on.

<br />

Now, let's read in the actual take and take a look at it.
</div>

In [2]:
df = pd.read_csv('./imdb.csv')
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

<div class="alert alert-block alert-warning">
Now, we're actually reading in our reviews scraped from IMDB's site. This data includes two columns: a review column including the actual text of the review, and a sentiment column including a String representing the sentiment of the review.
</div>

### Reformatting the Data for BERT

---

In [3]:
df['sentiment'] = (df['sentiment']=='positive').astype(int)
df.columns = ['text', 'label']
df.head()

text  label
0  One of the other reviewers has mentioned that ...      1
1  A wonderful little production. <br /><br />The...      1
2  I thought this was a wonderful way to spend ti...      1
3  Basically there's a family where a little boy ...      0
4  Petter Mattei's "Love in the Time of Money" is...      1

<div class="alert alert-block alert-warning">
Before inputting our data into the BERT model, we'll need to reformat our columns so the input is only numeric data, since the BERT model doesn't understand Strings or any text data.

<br />

So, we're just assigning 1 to the *positive* labels, and we're assigning 0 to the *negative* labels. Then, we'll rename our columns to make them slightly more readable.
</div>

### Prepare for BERT Training

---

In [4]:
df_train, df_valid = train_test_split(df, test_size=0.2)
len(df), len(df_train), len(df_valid)

(50000, 40000, 10000)

<div class="alert alert-block alert-warning">
Before I go any further into the specifics of model training, let me just make sure we all know why we want to train BERT or any other neural network. As a I said earlier, BERT was built to be used for many different NLP tasks, such as predicting next sentences, topic classification, sentiment classification, and many more NLP tasks. More technically, BERT was built for making predictions about sequential data, which many NLP tasks are. So, BERT is useful for us here, because we're interested in sentiment analysis.

<br />

Now, why do we want to train our network on more data, if BERT learned from millions and millions of sentences by Google for various NLP tasks already. Well, it's because BERT was trained on a lot of general data, so in order for us to get more accurate predictions, we need to fine tune our model to focus more on movie reviews, rather than make predictions on all Wikipedia data. Basically, BERT provides a great start for a lot of NLP tasks, but we will usually still need to fine tune it.

<br />
    
Now, normally when we want to train any model, we'll split our data into a training set and test set. The training set will be used for the model to actually learn from our training data, and the test set will be used for us to validate the accuracy of our model after training. The training set is usually somewhere between 70-90% of our entire dataset. For now, we'll just assign the size of our test set to be 20% of our total movie reviews.

<br />

Notice, when we print out the lengths of each dataset, the training set has 40K reviews, and our test set has 10K reviews.
</div>

In [5]:
hyperparameters = {
    'num_train_epochs': 3,
    'learning_rate': 1e-5
}

<div class="alert alert-block alert-warning">
Another thing we usually do before training our model is assigning some hyperparameters that are relevant to our model. These can be modified if we train our model multiple times to see which one gives us a better accuracy. It's best practice to save these hyperparameters and their resulting accuracies after we train our model with different hyperparameter settings.

<br />

Let's not worry to much about this step, but just know that the number of training epochs, learning rate, and many other hyperparameters can be set and help improve the accuracy of our model if we play around with these values.
</div>

### Training the BERT Model

---

In [6]:
model = ClassificationModel('bert', 'bert-large-cased', args=hyperparameters, use_cuda=False)

Some weights of the model checkpoint at bert-large-cased were not used when initializing BertForSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at 

<div class="alert alert-block alert-warning">
Now, in a single line, we're actually loading in the BERT model. Again, just in a single line, we're able to basically load in a very accurate model used by Google for so many NLP features internally and on their site. Notice, we're loading in our hyperparameter specifications as well. This *ClassificationModel* function can use many other NLP model for different classification tasks, such as GPT and many others.

<br />

So, the *bert* argument specifies that we want to use the BERT model for classification, rather than GPT or some other NLP model for classification. Then, the *bert-large-cased* is a tokenizer, that basically converts every single word from our reviews into a number. They basically have a vocabulary behind-the-scenes that performs a 1-to-1 mapping of a word in the vocabulary to some arbitrary number. We'll go into this a little deeper later.

<br />

There is also a *use_cuda* argument used for our function here, but just don't worry about this. Just know that we can specify other arguments in our function.
</div>

In [7]:
model.train_model(df_train)

/home/darius/.local/lib/python3.6/site-packages/simpletransformers/classification/classification_model.py:446: UserWarning: Dataframe headers not specified. Falling back to using column 0 as text and column 1 as labels.
  "Dataframe headers not specified. Falling back to using column 0 as text and column 1 as labels."


Process ForkPoolWorker-5:
Process ForkPoolWorker-1:
Process ForkPoolWorker-6:
Process ForkPoolWorker-4:
Process ForkPoolWorker-2:
Process ForkPoolWorker-3:
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/darius/.

KeyboardInterrupt: 

<div class="alert alert-block alert-warning">
Finally, we'll be able to train our model using the training data we split earlier. I'm not going to run this line of code now, because it would take hours to train our 40K lines of movie reviews to iterate until we get an extremely accurate model.

<br />

Luckily, I trained on this data before our talk. So, I'll just import the model I trained earlier, then we can start actually making predictions using our trained model. So, let me quickly import it now, so we don't have to just wait here for a few hours.
</div>

In [8]:
tokenizer = AutoTokenizer.from_pretrained("lvwerra/bert-imdb")
model = AutoModelForSequenceClassification.from_pretrained("lvwerra/bert-imdb")

<div class="alert alert-block alert-warning">
Here, we're just loading in the model I trained earlier, which I had named *bert-imdb*. Along with it, we're loading in the pre-trained tokenizer I mentioned earlier, which just maps the words in reviews to some fixed number.
</div>

### Predicting Sentiment with BERT

---

In [9]:
test = 'i love this movie'

<div class="alert alert-block alert-warning">
Now that we've done the more complicated stuff associated to model building and model training, let's actually get to the fun part, which is testing the accuracy of our model when we send in some new inputs our model hasn't seen before.

<br />

Here, let's say we're interested in seeing how our model classifies a review that simply says *i love this movie*. This is a very simple example, but in a little bit, I'll show you how our model can do pretty well against more complicated examples that you guys come up with. For now, we'll assign our review to a variable called *test*.
</div>

In [10]:
test_tensor = tokenizer.encode(test, return_tensors="pt")
test_tensor

tensor([[ 101,  178, 1567, 1142, 2523,  102]])

<div class="alert alert-block alert-warning">
Again, our BERT model can only understand numeric data. So, we'll just quickly tokenize our input data. Notice, our review contains 4 words, but the output of our tokenizer contains 6 tokens. At first you might think this is a mistake, but it's actually just a great example of how complex and smart our pre-trained tokenizer is. This tokenizer basically looks at some text, and doesn't just map individual words to a number. But, it actually maps parts of words Google deemed was important in understanding the context behind these sentences. So, 101 might refer to the word *I*, then 178 might refer to the word *love*, or 178 might refer to the first two letters of *love*: *lo*. We can take a closer look at the internals of what these numbers are actually mapped too, but unfortunately we'd need to look at the internals of the tokenizer object. So, for now, just take my word for it.

<br />

Anyways, all we're doing here is converting our review from a series of words to a series of mapped numbers.
</div>

In [11]:
outputs = model.forward(test_tensor)
outputs

SequenceClassifierOutput(loss=None, logits=tensor([[-3.6446,  4.2778]], grad_fn=<AddmmBackward>), hidden_states=None, attentions=None)

<div class="alert alert-block alert-warning">
Now, once we've converted our review to some numbers, we can finally see what our model thinks about the sentiment of this very basic review. To do this, we just call the *forward* function on our model, and we'll receive some output, including something called the loss, logits, and some other details. We're only interested in the logits here, so we're going to extract the logits from this output.
</div>

### Interpreting BERT Output

---

In [12]:
logits = outputs.logits
logits

tensor([[-3.6446,  4.2778]], grad_fn=<AddmmBackward>)

<div class="alert alert-block alert-warning">
So, that's all we're doing here. We're just grabbing the logits from the output. And, logits are basically just inputs of a softmax function in tensorflow. Meaning, they're some values we need in order to calculate the probabilities of the review being positive or negative.
</div>

In [13]:
probs = tf.nn.softmax(logits[0].detach())
probs

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([3.6240328e-04, 9.9963760e-01], dtype=float32)>

<div class="alert alert-block alert-warning">
Now, let's go ahead and input these into our softmax function, which will spit out these probabilities. I'm going to do one more step so that these probabilities are in a more readible format.
</div>

In [14]:
prob_neg = probs[0].numpy()
prob_neg

0.00036240328

In [15]:
prob_pos = probs[1].numpy()
prob_pos

0.9996376

<div class="alert alert-block alert-warning">
All I'm doing in this last step is reformatting these probabilities. Notice, our model is extremely certain that this review is positive, and not negative at all. Specifically, it's saying there's a 99% chance that this review is positive, which I think we can all agree that it is.
</div>

# Saving our Predictions to a Database

### Loading Sentiments in Database

---

In [1]:
from sqlalchemy import create_engine

In [2]:
db_string = 'postgres://postgres:6300FitchP(@localhost:5432/postgres'

In [3]:
db = create_engine(db_string)

/home/darius/.local/lib/python3.6/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


In [4]:
db.execute("CREATE TABLE IF NOT EXISTS reviews (review_id SERIAL PRIMARY KEY, review VARCHAR(200) NOT NULL, sentiment DECIMAL NOT NULL)")

OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 5432?

(Background on this error at: http://sqlalche.me/e/13/e3q8)

In [30]:
review_sql = "INSERT INTO reviews (review_id, review, sentiment) VALUES (1, {0}, {1})".replace(test, prob_pos)
db.execute(review_sql)

NameError: name 'test' is not defined

# Creating a Sentiment Service

- flask web service

# Creating a Sentiment Web Page

- plotly visualizations
- dash front-end

### References

---

- [History of NLP](https://www.dataversity.net/a-brief-history-of-natural-language-processing-nlp/)
- [Training BERT](https://lvwerra.github.io/trl/03-bert-imdb-training/)